ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/WindowStoreCacheBypass.java

```diff
@@ -106,59 +119,209 @@ private WindowStoreCacheBypass() {}
           return result;
         }
       } catch (final InvalidStateStoreException e) {
-        throw new InvalidStateStoreException(
-            "State store is not available anymore and may have been migrated to another instance; "
-                + "please re-discover its location from the state metadata.", e);
+        throw new InvalidStateStoreException(STORE_UNAVAILABLE_MESSAGE, e);
       }
     }
-    return new EmptyKeyValueIterator();
+    return new EmptyWindowStoreIterator();
   }
 
-  @SuppressWarnings("unchecked")
   private static WindowStoreIterator<ValueAndTimestamp<GenericRow>> fetchUncached(
       final ReadOnlyWindowStore<GenericKey, ValueAndTimestamp<GenericRow>> windowStore,
       final GenericKey key,
       final Instant lower,
       final Instant upper
   ) {
-    if (windowStore instanceof MeteredWindowStore) {
-      final StateSerdes<GenericKey, ValueAndTimestamp<GenericRow>> serdes;
+    if (!(windowStore instanceof MeteredWindowStore)) {
+      throw new IllegalStateException("Expecting a MeteredWindowStore");
+    }
+    final StateSerdes<GenericKey, ValueAndTimestamp<GenericRow>> serdes = getSerdes(windowStore);
+    final WindowStore<Bytes, byte[]> wrapped = getInnermostStore(windowStore);
+    final Bytes rawKey = Bytes.wrap(serdes.rawKey(key));
+    final WindowStoreIterator<byte[]> fetch = wrapped.fetch(rawKey, lower, upper);
+    return new DeserializingIterator(fetch, serdes);
+  }
+
+  public static KeyValueIterator<Windowed<GenericKey>, ValueAndTimestamp<GenericRow>> fetchRange(
+          final ReadOnlyWindowStore<GenericKey, ValueAndTimestamp<GenericRow>> store,
+          final GenericKey keyFrom,
+          final GenericKey keyTo,
+          final Instant lower,
+          final Instant upper
+  ) {
+    Objects.requireNonNull(keyFrom, "lower key can't be null");
+    Objects.requireNonNull(keyTo, "upper key can't be null");
+    final List<ReadOnlyWindowStore<GenericKey, ValueAndTimestamp<GenericRow>>> stores
+            = getStores(store);
+    for (final ReadOnlyWindowStore<GenericKey, ValueAndTimestamp<GenericRow>> windowStore :stores) {
```

Comment:
You can probably put this for loop in a method that takes a lambda:

```
findFirstNonEmptyIterator(... stores, Function<ReadOnlyWindowStore<GenericKey, ValueAndTimestamp<GenericRow>>, KeyValueIterator<Windowed<GenericKey>> func) {
  for (...) {
  
  }
}
```
You might have to make a generic parameter of the method if you want to handle the WindowStoreIterator as well, but it could simplify the code a bit more.

Reply:
I have tried to implement that, but it required a bit of type casting and I am nervous about that